cmd/kubeapps-apis/plugins/helm/packages/v1alpha1/server.go

```diff
@@ -822,70 +826,106 @@ func (s *Server) CreateInstalledPackage(ctx context.Context, request *corev1.Cre
-	appRepo, caCertSecret, authSecret, err := s.getAppRepoAndRelatedSecrets(ctx, repoName, repoNamespace)
+	// Create an action config for the target namespace.
+	actionConfig, err := s.actionConfigGetter(ctx, request.GetTargetContext())
 	if err != nil {
-		return nil, status.Errorf(codes.Internal, "Unable to fetch app repo %q from namespace %q: %v", repoName, repoNamespace, err)
+		return nil, status.Errorf(codes.Internal, "Unable to create Helm action config: %v", err)
 	}
 
-	userAgentString := fmt.Sprintf("%s/%s/%s/%s", UserAgentPrefix, pluginDetail.Name, pluginDetail.Version, version)
-
-	log.Infof("fetching chart %q with user-agent %q", chartID, userAgentString)
-
-	// Grab the chart itself
-	ch, err := handlerutil.GetChart(
-		&chart.Details{
-			AppRepositoryResourceName:      appRepo.Name,
-			AppRepositoryResourceNamespace: appRepo.Namespace,
-			ChartName:                      chartName,
-			Version:                        chartVersion,
-		},
-		appRepo,
-		caCertSecret, authSecret,
-		s.chartClientFactory.New(appRepo.Spec.Type, userAgentString),
-	)
+	release, err := agent.CreateRelease(actionConfig, request.GetName(), request.GetTargetContext().GetNamespace(), request.GetValues(), ch, registrySecrets)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "Unable to create helm release %q in the namespace %q: %v", request.GetName(), request.GetTargetContext().GetNamespace(), err)
+	}
 
-	// Create an action config for the target namespace.
-	namespace := request.GetTargetContext().GetNamespace()
 	cluster := request.GetTargetContext().GetCluster()
 	if cluster == "" {
 		cluster = s.globalPackagingCluster
 	}
-	actionConfig, err := s.actionConfigGetter(ctx, cluster, namespace)
+	return &corev1.CreateInstalledPackageResponse{
+		InstalledPackageRef: &corev1.InstalledPackageReference{
+			Context: &corev1.Context{
+				Cluster:   cluster,
+				Namespace: release.Namespace,
+			},
+			Identifier: release.Name,
+			Plugin:     GetPluginDetail(),
+		},
+	}, nil
+}
+
+// UpdateInstalledPackage updates an installed package.
+func (s *Server) UpdateInstalledPackage(ctx context.Context, request *corev1.UpdateInstalledPackageRequest) (*corev1.UpdateInstalledPackageResponse, error) {
+	installedRef := request.GetInstalledPackageRef()
+	releaseName := installedRef.GetIdentifier()
+	contextMsg := fmt.Sprintf("(cluster=%q, namespace=%q)", installedRef.GetContext().GetCluster(), installedRef.GetContext().GetNamespace())
+	log.Infof("+helm UpdateInstalledPackage %s", contextMsg)
+
+	// Determine the chart used for this installed package.
+	// We may want to include the AvailablePackageRef in the request, given
+	// that it can be ambiguous, but we dont yet have a UI that allows the
+	// user to select which chart to use, so until then.
+	detailResponse, err := s.GetInstalledPackageDetail(ctx, &corev1.GetInstalledPackageDetailRequest{
+		InstalledPackageRef: installedRef,
+	})
```

Comment:
Don't we? I guess we can extend the existing `dashboard/src/components/SelectRepoForm/SelectRepoForm.tsx` to support more scenarios like this.

Reply:
Yeah, it'd also be nice to send it explicitly (given it's already returned to the client with the detail) since we then wouldn't need to fetch the detail here again (took that as a shortcut, could have just fetched the related available pkg ref, and probably should... I'll add a TODO). Greg and I weren't 100% whether we should include the available pkg reference... see the [thread on his PR adding the proto messages](https://github.com/kubeapps/kubeapps/pull/3425#discussion_r709712844) for more context.